text/0000-target-bundles.md

```diff
@@ -0,0 +1,245 @@
+Crockford himself:
+
+> Suppose you are using JSON to keep configuration files, which you would like to annotate. Go
+> ahead and insert all the comments you like. Then pipe it through JSMin before handing it to your
+> JSON parser.
+
+This RFC proposes to use similar scheme: adjust the build system to remove comments from all the
+JSON target specifications before packaging them for use by rustc. This way we get to keep using
+JSON and can keep on having comments in the checked-in target specifications.
+
+Similar preprocessing step could be used to implement some form of target inheritance so the
+duplication between built-in targets could be reduced greatly.
+
+[comments]: https://github.com/rust-lang/rust/pull/34980#issuecomment-234683183
+[jsmin]: https://plus.google.com/+DouglasCrockfordEsq/posts/RK8qyGVaGSr
+
+### Target quirks and cfg
+
+Currently we have a few keys dealing with the configuration variables only: `target_os`,
+`target_env`, `target_vendor`, `target_family`, `target_endian`, `target_pointer_width`¹, etc;
+Then there’s options from which some configuration variables are derived, but they are also used to
+tweak compilation: `is_like_windows`, `is_like_osx`, `is_like_solaris`, etc.
+
+This RFC proposes replacing these keys with a different set of keys which explicitly control
+configuration variables and keys which control compilation details:
+
+* `cfg: {"target_env": "msvc", windows: null }` would result in two cofiguration variables
+`#[cfg(windows)]` and `#[cfg(target_env = "msvc")]` evaluating to true, but no variables provided
+by the `cfg` key would get used to influence the behaviour of the compiler; and
+* `debuginfo: ["CodeView", 1]` some targets require different debuginfo format than what LLVM
+generates by default. MSVC targets want CodeView version 1, OS X and Android want Dwarf version 2,
+while LLVM appears to use the highest supported version of Dwarf by default.
+* similarly for many other variables which tweak the way compilation is done.
+
+¹: Technically, `target_pointer_width` is used in trans, but it does not provide any extra
+information over `data_layout`, which is what it should be using.
+
+This would allow selectively reusing various conditional implementations that are present in the
+compiler for a custom target, when none of the `is_like_*` variables would fully suit the target.
+Moreover, being able to specify arbitrary cfg variables would allow easily adapting for various
+miniscule details related to the targets. For example, the targets for ARM CPUs with NEON support
+could export `target_has_neon` without any extra language support.
+
+That being said, it might make sense to have a whitelist of stable options and cfg variables and
+keep everything else unstable for some duration.
+
+### Proposed JSON key-values
+
+```js
+{
+    // REQUIRED
+    "llvm-target": "x86_64-unknown-linux-gnu",              // LLVM target triple (does not need to match with rustc triple)
```

Comment:
Don't use dashes and instead use underscores so that you can just use `#[derive(RustcEncodable, RustcDecodable)]`. The dashes are the problem as I pointed out in https://github.com/rust-lang/rust/pull/32988